### PR TITLE
[docs] refactor docs for easier info parsing

### PR DESCRIPTION
**Pull request by sayakpaul**
Approach taken:

* Keep the README as less cluttered as possible. 
* Add docs for each model. Only when we have more training algos can we include separate sections within model docs.
* Once we have I2V support for a model that should go into the respective model accordingly.
* Have a central doc for memory optimizations.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -128,540 +125,24 @@ eval $cmd
 echo -ne "-------------------- Finished executing script --------------------\n\n"
 ```
 
-### Inference:
```
Comment (sayakpaul) [inline README.md:131]: Went to specific model docs.

File: README.md [hunk 2]
```diff
@@ -128,540 +125,24 @@ eval $cmd
...
 
-If you would like to use a custom dataset, refer to the dataset preparation guide [here](./assets/dataset.md).
-
-> [!NOTE]
-> To lower memory requirements:
-> - Use a DeepSpeed config to launch training (refer to [`accelerate_configs/deepspeed.yaml`](./accelerate_configs/deepspeed.yaml) as an example).
-> - Pass `--precompute_conditions` when launching training.
-> - Pass `--gradient_checkpointing` when launching training.
-> - Pass `--use_8bit_bnb` when launching training. Note that this is only applicable to Adam and AdamW optimizers.
-> - Do not perform validation/testing. This saves a significant amount of memory, which can be used to focus solely on training if you're on smaller VRAM GPUs.
```
Comment (sayakpaul) [inline README.md:398]: In `docs/training/optimizations.md`.

File: README.md [hunk 3]
```diff
@@ -128,540 +125,24 @@ eval $cmd
...
-
-## Memory requirements
```
Comment (sayakpaul) [inline README.md:400]: We don't need as have it here already:
https://github.com/a-r-r-o-w/finetrainers/tree/main/training

File: docs/training/cog.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
+## Training
```
Comment (a-r-r-o-w) [inline docs/training/cog.md:1]: Would rename this file to cogvideox.md

File: docs/training/cog.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
```
Comment (a-r-r-o-w) [inline docs/training/cog.md:2]: Could you add the memory numbers to keep it consistent with other model docs? Will clean this up a bit more later but for now should be good

```
Training configuration: {
    "trainable parameters": 132120576,
    "total samples": 69,
    "train epochs": 1,
    "train steps": 10,
    "batches per device": 1,
    "total batches observed per epoch": 69,
    "train batch size": 1,
    "gradient accumulation steps": 1
}

| stage                         | memory_allocated  | max_memory_reserved |
|:-----------------------------:|:-----------------:|:-------------------:|
| after precomputing conditions |  8.880            | 8.941               |
| after precomputing latents    |  9.300            | 12.441              |
| before training start         | 10.622            | 20.701              |
| after epoch 1                 | 11.145            | 20.701              |
| before validation start       | 11.145            | 20.702              |
| after validation end          | 11.145            | 28.324              |
| after training end            | 11.144            | 11.592              |
```

File: docs/training/cog.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+Assuming your LoRA is saved and pushed to the HF Hub, and named `my-awesome-name/my-awesome-lora`, we can now use the finetuned model for inference:
+
+```diff
+import torch
+from diffusers import DiffusionPipeline
+from diffusers.utils import export_to_video
+
+pipe = DiffusionPipeline.from_pretrained(
+    "THUDM/CogVideoX-2b", torch_dtype=torch.float16
```
Comment (a-r-r-o-w) [inline docs/training/cog.md:96]:
```suggestion
    "THUDM/CogVideoX-5b", torch_dtype=torch.bfloat16
```

I changed the defaults in the training script to CogVideoX-5b. This is because it is very hard to get anything good out of 2b and it has very low usage among finetuners IMO

File: docs/training/cog.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+).to("cuda")
++ pipe.load_lora_weights("my-awesome-name/my-awesome-lora", adapter_name="ltxv-lora")
```
Comment (a-r-r-o-w) [inline docs/training/cog.md:98]:
```suggestion
+ pipe.load_lora_weights("my-awesome-name/my-awesome-lora", adapter_name="cogvideox-lora")
```

File: docs/training/cog.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+echo -ne "-------------------- Finished executing script --------------------\n\n"
+```
+
+## Inference
+
+Assuming your LoRA is saved and pushed to the HF Hub, and named `my-awesome-name/my-awesome-lora`, we can now use the finetuned model for inference:
+
+```diff
+import torch
+from diffusers import DiffusionPipeline
```
Comment (a-r-r-o-w) [inline docs/training/cog.md:92]:
```suggestion
from diffusers import CogVideoXPipeline
```

File: docs/training/cog.md [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+from diffusers.utils import export_to_video
+
+pipe = DiffusionPipeline.from_pretrained(
```
Comment (a-r-r-o-w) [inline docs/training/cog.md:95]:
```suggestion
pipe = CogVideoXPipeline.from_pretrained(
```

File: docs/training/cog.md [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    "THUDM/CogVideoX-2b", torch_dtype=torch.float16
+).to("cuda")
++ pipe.load_lora_weights("my-awesome-name/my-awesome-lora", adapter_name="ltxv-lora")
++ pipe.set_adapters(["ltxv-lora"], [0.75])
```
Comment (a-r-r-o-w) [inline docs/training/cog.md:99]:
```suggestion
+ pipe.set_adapters(["cogvideox-lora"], [0.75])
```

File: docs/training/cog.md (hunk 5 above)
Comment (a-r-r-o-w) [reply docs/training/cog.md:2]: This is with the following settings btw:

```
LoRA with rank 128, batch size 1, gradient checkpointing, optimizer adamw, `49x480x720` resolutions, **with precomputation**:
```

File: docs/training/hunyuan.md [hunk 11]
```diff
@@ -0,0 +1,175 @@
+# Training
```
Comment (a-r-r-o-w) [inline docs/training/hunyuan.md:1]: Would rename this file to `hunyuan_video.md` just to keep consistent with actual model names

File: docs/training/ltx.md [hunk 12]
```diff
@@ -0,0 +1,163 @@
+## Training
```
Comment (a-r-r-o-w) [inline docs/training/ltx.md:1]: Would rename this file to `ltx_video.md` for same reason as above

Comment (a-r-r-o-w) [reply docs/training/ltx.md:1]: For all the model files, would maybe be good to add a `# {model_name}` header too

File: README.md [hunk 13]
```diff
@@ -120,548 +112,43 @@ cmd="accelerate launch --config_file accelerate_configs/uncompiled_2.yaml --gpu_
...
-| after validation end    | 39.748           | 58.385              |
-| after epoch 1           | 39.748           | 40.910              |
-| after training end      | 25.288           | 40.910              |
+## Notes
 
-Note: requires about `59` GB of VRAM when validation is performed.
+* The example training commands have the below thoughts in mind to prioritize memory-optimized runs:
+  * don't contain any arguments to enable validation inference.
+  * gradient checkpointing and gradient accumulation enabled.
+  * precompuation enabled.
```
Comment (a-r-r-o-w) [inline README.md:149]: I think okay to remove this part from main README

File: README.md [hunk 14]
```diff
@@ -120,548 +112,43 @@ cmd="accelerate launch --config_file accelerate_configs/uncompiled_2.yaml --gpu_
...
-output = pipe(
-    prompt="A cat walks on the grass, realistic",
-    height=320,
-    width=512,
-    num_frames=61,
-    num_inference_steps=30,
-).frames[0]
-export_to_video(output, "output.mp4", fps=15)
-```
+| Model Name | Tasks                | Ckpts Tested                                                                                                                                         | Min. GPU<br>VRAM | Comments                                                     |
```
Comment (a-r-r-o-w) [inline README.md:132]: Would remove the comments column here tbh. All models are supported with multi-resolution and multi-frames so far.

LTX is fast to train but it is extremely hard to teach it new styles tbh. I have had very limited success getting a good lora. Have not figured out the best settings yet but continuing to try.

**Review by a-r-r-o-w — COMMENTED** [review]
Thanks for the incredible refactor!

File: README.md (hunk 14 above)
Comment (sayakpaul) [reply README.md:132]: Oh I kept the comments column for including anything we should note that couldn't be fit in the other columns. Anything that's out of the ordinary and is absolutely important for the users to know. 

WDYT?

File: docs/training/cog.md (hunk 5 above)
Comment (sayakpaul) [reply docs/training/cog.md]: Thanks! I just added it in 1f81197. Could you also let me know about the ckpt you used for this? Is it 1.0 5B or 1.5 5B?

Comment (sayakpaul) [reply docs/training/cog.md]: I think this is 5b. So, updated accordingly.

**Comment by sayakpaul** [comment]
@a-r-r-o-w thanks for the reviews! In the latest commits:

* I have addressed all your comments and left a question for you [here](https://github.com/a-r-r-o-w/finetrainers/pull/175#discussion_r1903263762). 
* Introduced an ultra-short section from the `docs/training/README.md` to reference the model-specific docs for easier navigation.

**Review by a-r-r-o-w — APPROVED** [review]
thanks!
